source/apiVolontaria/apiVolontaria/views.py

```diff
@@ -257,3 +264,130 @@ def post(self, request):
                 {'detail': error},
                 status=status.HTTP_500_INTERNAL_SERVER_ERROR
             )
+
+
+class ForgotPassword(APIView):
+    """
+    post:
+    Create a new token allowing user to change his password.
+    """
+    permission_classes = ()
+    authentication_classes = ()
+
+    def post(self, request, *args, **kwargs):
+        if settings.CONSTANT['EMAIL_SERVICE'] is not True:
+            # Without email this functionality is not provided
+            return Response(status=status.HTTP_501_NOT_IMPLEMENTED)
+
+        # get user from the email given in data
+        user = User.objects.get(username=request.data["username"])
+
+        # remove old tokens to change password
+        tokens = ActionToken.objects.filter(
+            type='password_change',
+            user=user,
+        )
+
+        for token in tokens:
+            token.expire()
+
+        # create the new token
+        token = ActionToken.objects.create(
+            type='password_change',
+            user=user,
+        )
+
+        # Send the new token by e-mail to the user
+        MAIL_SERVICE = settings.SETTINGS_IMAILING
+        FRONTEND_SETTINGS = settings.CONSTANT['FRONTEND_INTEGRATION']
+
+        button_url = FRONTEND_SETTINGS['FORGOT_PASSWORD_URL'].replace(
+            "{{token}}",
+            str(token)
+        )
+
+        email = IMailing.create_instance(
+            MAIL_SERVICE["SERVICE"],
+            MAIL_SERVICE["API_KEY"],
+        )
+
+        response_send_mail = email.send_templated_email(
+            email_from=MAIL_SERVICE["EMAIL_FROM"],
+            template_id=MAIL_SERVICE["TEMPLATES"]["FORGOT_PASSWORD"],
+            list_to=[user.email],
+            context={
+                "forgot_password_url": button_url,
+            },
+        )
+
+        if response_send_mail["code"] == "failure":
+            content = {
+                'detail': _("Your token has been created but no email "
+                            "has been sent. Please contact the "
+                            "administration."),
+            }
+            return Response(content, status=status.HTTP_201_CREATED)
+
+        else:
+            return Response(status=status.HTTP_201_CREATED)
+
+
+class ChangePassword(APIView):
+    """
+    post:
+    Get a token and a new password and change the password of
+    the token's owner.
+    """
+    authentication_classes = ()
+    permission_classes = ()
+    serializer_class = serializers.UserBasicSerializer
+
+    def post(self, request):
+        token = request.data.get('token')
```

Comment:
What happens if I don't provide a token? Should we add a serializer to be sure we get required parameters?

Reply:
Actually we have an `HTTP_400_BAD_REQUEST` with this message :

```
{'detail': "\"None\" is not a valid token."}
```
I think it can be good to add a serializer as you propose.

Reply:
Should we implement it now?